src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,429 @@
+
+                    if (activeProcessors != _totalProcessorsRunning)
+                    {
+                        _logger.LogTrace($"Detected that number of processors equals {activeProcessors} but this processor is running with {_totalProcessorsRunning} known processors");
+                        _logger.LogTrace("Restarting and recomputing");
+                        cts.Cancel();
+                    }
+
+                    await Task.Delay(30000, cts.Token);
+                }
+            });
+
+            return Task.CompletedTask;
+        }
+
+        private Task RenewOwnership(CancellationToken ct)
+        {
+            Task.Run(async () =>
+            {
+                while (!ct.IsCancellationRequested)
+                {
+                    foreach (var partition in _ownedPartitions)
+                    {
+                        _logger.LogTrace($"Renewing ownership for partition {partition}");
+                        BlobClient ownershipClient = _containerClient.GetBlobClient($"ownedpartitions/{partition}");
+                        var leaseClient = ownershipClient.GetBlobLeaseClient(_processorId);
+                        await leaseClient.RenewAsync(default, ct);
+                    }
+
+                    await Task.Delay(20000);
+                }
+            });
+
+            return Task.CompletedTask;
+        }
+
+        private async Task StartAssignedPartitionProcessor(CancellationToken ct, List<int> ownedPartitions)
+        {
+            var maximumBatchSize = 100;
+
+            string[] partitions = new string[ownedPartitions.Count];
+            for (var i = 0; i < ownedPartitions.Count; i++)
+            {
+                partitions[i] = ownedPartitions.ElementAt(i).ToString();
+            }
+
+            var eventHubName = EnsureArg.IsNotNull(_eventHubClientOptions.EventHubName, nameof(_eventHubClientOptions.EventHubName));
+            var eventHubNamespaceFQDN = EnsureArg.IsNotNull(_eventHubClientOptions.EventHubNamespaceFQDN, nameof(_eventHubClientOptions.EventHubNamespaceFQDN));
+            var eventHubConsumerGroup = EnsureArg.IsNotNull(_eventHubClientOptions.EventHubConsumerGroup, nameof(_eventHubClientOptions.EventHubConsumerGroup));
+            var tokenCrendential = EnsureArg.IsNotNull(_eventHubClientOptions.EventHubTokenCredential, nameof(_eventHubClientOptions.EventHubTokenCredential));
+
+            var processor = new AssignedPartitionProcessor(
+                _eventBatchingService,
+                _checkpointClient,
+                _logger,
+                partitions,
+                maximumBatchSize,
+                eventHubConsumerGroup,
+                tokenCrendential,
+                eventHubName,
+                eventHubNamespaceFQDN);
+
+            try
+            {
+                await processor.StartProcessingAsync(ct);
+                await Task.Delay(Timeout.Infinite, ct);
```

Comment:
added a flowchart here: https://github.com/microsoft/iomt-fhir/blob/personal/wiyochum/partition-locking-prototype/docs/PartitionLocking.md but essentially when the AssignedPartitionProcessor starts we will cache off the number of known processors at the time. Every 30 seconds when we register the current processor we also check the storage account to see if the current number of total processors matches the number of processors that we have cached. If the number of processors differs from the cache then we restart.

Reply:
Thank you for the flowchart.  My main concern with this approach is if there is any change in ownership, everything restarts, likely other nodes as well unless the a processor can claim ownership before the others detect it.  That restart will include the entire loop evaluation as well.  It seems like we could get into a scenario where if things get slightly off, loss of ownership could cascade continuously and we never get into a state where all partitions are claimed.

Thoughts on explicitly tracking the partitions owned and assigning them their own processor and either canceling that processor when ownership is lost or adding a new one when a partition is claimed?

Also, if you have a link handy for the original event hub logic it would be useful to compare to.

Reply:
Looking at the logic again, it is based solely on the number of processors, not a partition losing its lease so that seems more stable but seems likely for high churn, especially in scenarios where K8 is adding a new pod before taking one down.

Reply:
One other side effect of having dedicated processor instances per partition is we would no longer be able to dynamically scale the buffer size up as number of partitions in the processor decreases.  We would have to look at other solutions if we make the change I suggested.

Reply:
This is the link to the original event hub logic which steals partitions from other processors but does not immediately notify the other processors that the partition has been stolen, which leads to two processors having the same data in memory and writing the same data to the fhir service

https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/Processor/PartitionLoadBalancer.cs#L412